aad/spring-cloud-azure-starter-active-directory/aad-web-application/src/main/java/com/azure/spring/sample/aad/security/AadConditionalAccessFilter.java

```diff
@@ -0,0 +1,93 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.spring.sample.aad.security;
+
+import com.azure.spring.cloud.autoconfigure.aad.implementation.constants.Constants;
+import com.azure.spring.cloud.autoconfigure.aad.implementation.oauth2.AadClientRegistrationRepository;
+import org.springframework.http.HttpHeaders;
+import org.springframework.security.oauth2.client.ClientAuthorizationRequiredException;
+import org.springframework.util.StringUtils;
+import org.springframework.web.filter.OncePerRequestFilter;
+import org.springframework.web.reactive.function.client.WebClientResponseException;
+
+import javax.servlet.FilterChain;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * Handle the {@link WebClientResponseException} in On-Behalf-Of flow.
+ *
+ * <p>
+ * When the resource-server needs re-acquire token(The request requires higher privileges than provided by the access
+ * token in On-Behalf-Of flow.), it can send a 403 with information in the WWW-Authenticate header to web client ,web
+ * client will throw {@link WebClientResponseException}, web-application can handle this exception to challenge the
+ * user.
+ *
+ * @see OncePerRequestFilter
+ */
+public class AadConditionalAccessFilter extends OncePerRequestFilter {
+
+    /**
+     * Do filter.
+     *
+     * @param request the HttpServletRequest
+     * @param response the HttpServletResponse
+     * @param filterChain the FilterChain
+     * @throws IOException if an I/O related error has occurred during the processing
+     * @throws ServletException if an exception has occurred that interferes with the
+     *                          filterChain's normal operation
+     */
+    @Override
+    protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response,
+                                    FilterChain filterChain) throws IOException, ServletException {
+        // Handle conditional access policy, step 2.
```

Comment:
where are other steps?

Reply:
It's in spring-cloud-azure-autoconfigure.